adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -702,7 +702,23 @@ private Span newSpan(final String operationName, final Device authenticatedDevic
                                 command.getPayloadSize(),
                                 timer);
                     } else {
-                        onCommandReceived(sender, commandContext);
+                        getTenantConfiguration(sourceAddress.getTenantId(), commandContext.getTracingContext())
+                                .compose(tenantObject -> checkMessageLimit(tenantObject, command.getPayloadSize()))
+                                .compose(success -> {
+                                    onCommandReceived(sender, commandContext);
+                                    return Future.succeededFuture();
+                                })
+                                .otherwise(t -> {
+                                    // issue credit so that application(s) can send the next command
+                                    commandContext.reject(getErrorCondition(t), 1);
+                                    metrics.reportCommand(
+                                            command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                                            sourceAddress.getTenantId(),
+                                            ProcessingOutcome.from(t),
+                                            command.getPayloadSize(),
```

Comment:
so, if the message limit is reached, we do not process the command but bill the customer for it?

Reply:
In the above case, if the message limit is reached, the command is not processed and the corresponding metrics for that command is reported with `ProcessingOutcome` as `UNPROCESSABLE`. 

The metrics data from the backend having `ProcessingOutcome` tagged as `FORWARDED` or `UNPROCESSABLE` are used for the message limit calculation. The same is also applicable to _telemetry/event_ messages. What do you mean by bill the customer? Can you pls explain more?

Reply:
hmm. Let's say I have a limit of 10 MB. From my application I send several commands with a big payload which cannot be processed because I never include a subject. My understanding is that the commands will all *eat up* my message volume, i.e. I have to pay for them, regardless of whether they have reached the destination (device). So, once the limit is reached, will the payload still be accounted for, i.e. will I still need to pay for it, even if now all messages are being rejected? I think I should, because my sending the of these (feasible) commands still generates costs on our side, right? So, based on that, reporting the command's payload seems to be correct but should also be subjected to the *minimum message size* calculation, right?
The problem is that we cannot prevent the application from sending additional commands, thus causing more and more costs, because we only check the limit *after* we have accepted the messages via the AMQP Network ...

Reply:
> So, based on that, reporting the command's payload seems to be correct but should also be subjected to the _minimum message size_ calculation, right?

yes. This PR is to address the issue #1284, which is to equip the adapters with message limit checks while processing commands. The minimum message size calculation as described in issue #1330 will be part of another [PR](https://github.com/eclipse/hono/pull/1333).

> The problem is that we cannot prevent the application from sending additional commands, thus causing more and more costs, because we only check the limit _after_ we have accepted the messages via the AMQP Network ...

yes. I think the same is also applicable for the _Telemetry/Event_ messages i.e we cannot prevent the devices from sending additional messages after the limit is exceeded. If this is the indented behaviour i.e the customer will be charged for the additional non processed messages then it should be well documented, so that customers can design their implementation accordingly.  

Thereby making it also important in Hono to inform the devices and north bound applications with this exact error reason so that they can act upon it. With HTTP and AMQP protocols I believe that they are well informed with status codes _429_ and `amqp:resource-limit-exceeded`. But the MQTT protocol doesn't provide space for such an option and we are closing the MQTT connection as recommended by the spec. IMHO it would be worth to think in a separate issue how to communicate the errors effectively in case of MQTT. May be a separate topic to which the interested devices can subscribe to know about the errors until we support MQTT 5 ;-). MQTT 5 has option to send reason codes along with control packets.